File: go/overdrive_e2e_test.go
package overdrive_test

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"testing"

	overdrive "github.com/ALL-FOR-ONE-TECH/overdrive-db-go"
)

func tmpDB(t *testing.T, name string) string {
	t.Helper()
	dir := filepath.Join(os.TempDir(), "overdrive_sdk_e2e_go")
	os.MkdirAll(dir, 0755)
	p := filepath.Join(dir, name+".odb")
	os.Remove(p)
	os.Remove(p + ".wal")
	t.Cleanup(func() { os.Remove(p); os.Remove(p + ".wal") })
	return p
}

// TEST 1
func TestOpen_CreatesFile(t *testing.T) {
	p := tmpDB(t, "t01_open")
	odb, err := overdrive.Open(p)
	if err != nil { t.Fatalf("open: %v", err) }
	odb.Close()
	info, err := os.Stat(p)
	if err != nil { t.Fatal("❌ .odb file not created") }
	if info.Size() == 0 { t.Fatal("❌ .odb is 0 bytes") }
	fmt.Printf("✅ TEST 1 — .odb created (%d bytes)\n", info.Size())
}

// TEST 2
func TestInsertGet_Roundtrip(t *testing.T) {
	p := tmpDB(t, "t02_crud")
	odb, _ := overdrive.Open(p)
	defer odb.Close()
	odb.CreateTable("users")
	id, err := odb.Insert("users", map[string]any{"name": "Karthikeyan", "role": "engineer", "age": 28})
	if err != nil || id == "" { t.Fatalf("insert: %v", err) }
	doc, err := odb.Get("users", id)
	if err != nil || doc == nil { t.Fatalf("get: %v", err) }
	if doc["name"] != "Karthikeyan" { t.Fatalf("❌ name mismatch: %v", doc["name"]) }
	fmt.Printf("✅ TEST 2 — odb.Insert+Get: _id=%s, name=%v\n", id, doc["name"])
}

// TEST 3
func TestCount_Accurate(t *testing.T) {
	p := tmpDB(t, "t03_count")
	odb, _ := overdrive.Open(p)
	defer odb.Close()
	odb.CreateTable("items")
	n, _ := odb.Count("items")
	if n != 0 { t.Fatalf("❌ expected 0, got %d", n) }
	odb.Insert("items", map[string]any{"n": "A"})
	odb.Insert("items", map[string]any{"n": "B"})
	odb.Insert("items", map[string]any{"n": "C"})
	n, _ = odb.Count("items")
	if n != 3 { t.Fatalf("❌ expected 3, got %d", n) }
	fmt.Printf("✅ TEST 3 — odb.Count=3\n")
}

// TEST 4
func TestMultiGet_Fields(t *testing.T) {
	p := tmpDB(t, "t04_multi")
	odb, _ := overdrive.Open(p)
	defer odb.Close()
	odb.CreateTable("products")
	id1, _ := odb.Insert("products", map[string]any{"name": "Apple",  "price": 10})
	id2, _ := odb.Insert("products", map[string]any{"name": "Banana", "price": 5})
	id3, _ := odb.Insert("products", map[string]any{"name": "Cherry", "price": 25})
	d1, _ := odb.Get("products", id1)
	d2, _ := odb.Get("products", id2)
	d3, _ := odb.Get("products", id3)
	if d1["name"] != "Apple" || d2["name"] != "Banana" || d3["name"] != "Cherry" {
		t.Fatal("❌ field mismatch")
	}
	fmt.Printf("✅ TEST 4 — odb.Get per-doc fields correct\n")
}

// TEST 5
func TestUpdate_ChangesField(t *testing.T) {
	p := tmpDB(t, "t05_update")
	odb, _ := overdrive.Open(p)
	defer odb.Close()
	odb.CreateTable("cfg")
	id, _ := odb.Insert("cfg", map[string]any{"key": "theme", "val": "light"})
	ok, err := odb.Update("cfg", id, map[string]any{"val": "dark"})
	if err != nil || !ok { t.Fatalf("update failed: %v", err) }
	doc, _ := odb.Get("cfg", id)
	if doc["val"] != "dark" { t.Fatalf("❌ expected dark, got %v", doc["val"]) }
	fmt.Printf("✅ TEST 5 — odb.Update: light→dark\n")
}

// TEST 6
func TestDelete_RemovesDoc(t *testing.T) {
	p := tmpDB(t, "t06_delete")
	odb, _ := overdrive.Open(p)
	defer odb.Close()
	odb.CreateTable("logs")
	id1, _ := odb.Insert("logs", map[string]any{"msg": "e1"})
	id2, _ := odb.Insert("logs", map[string]any{"msg": "e2"})
	ok, _ := odb.Delete("logs", id1)
	if !ok { t.Fatal("❌ delete must return true") }
	n, _ := odb.Count("logs")
	if n != 1 { t.Fatalf("❌ count must be 1, got %d", n) }
	d, _ := odb.Get("logs", id1)
	if d != nil { t.Fatal("❌ deleted doc must be nil") }
	d2, _ := odb.Get("logs", id2)
	if d2 == nil { t.Fatal("❌ remaining doc must exist") }
	fmt.Printf("✅ TEST 6 — odb.Delete: count=1, deleted=nil\n")
}

// TEST 7
func TestPersist_AfterReopen(t *testing.T) {
	p := tmpDB(t, "t07_persist")
	var storedID string
	{
		odb, _ := overdrive.Open(p)
		odb.CreateTable("sessions")
		storedID, _ = odb.Insert("sessions", map[string]any{"token": "abc123", "user": "afot"})
		odb.Sync()
		odb.Close()
	}
	{
		odb, _ := overdrive.Open(p)
		defer odb.Close()
		n, _ := odb.Count("sessions")
		if n != 1 { t.Fatalf("❌ data must persist, count=%d", n) }
		doc, _ := odb.Get("sessions", storedID)
		if doc == nil { t.Fatal("❌ doc must exist after reopen") }
		if doc["token"] != "abc123" { t.Fatalf("❌ token mismatch: %v", doc["token"]) }
		fmt.Printf("✅ TEST 7 — persisted: token=%v, _id=%s\n", doc["token"], storedID)
	}
}

// TEST 8
func TestInsertBatch(t *testing.T) {
	p := tmpDB(t, "t08_batch")
	odb, _ := overdrive.Open(p)
	defer odb.Close()
	odb.CreateTable("orders")
	ids, err := odb.InsertBatch("orders", []map[string]any{
		{"order_id": "ORD-001", "amount": 150},
		{"order_id": "ORD-002", "amount": 200},
		{"order_id": "ORD-003", "amount": 75},
	})
	if err != nil || len(ids) != 3 { t.Fatalf("insertBatch: %v", err) }
	n, _ := odb.Count("orders")
	if n != 3 { t.Fatalf("❌ count must be 3, got %d", n) }
	for _, id := range ids {
		doc, _ := odb.Get("orders", id)
		if doc == nil { t.Fatalf("❌ inserted doc %s not found", id) }
	}
	fmt.Printf("✅ TEST 8 — odb.InsertBatch: %v\n", ids)
}

// TEST 9
func TestTableExists(t *testing.T) {
	p := tmpDB(t, "t09_tables")
	odb, _ := overdrive.Open(p)
	defer odb.Close()
	if odb.TableExists("ghost") { t.Fatal("❌ ghost should not exist") }
	odb.CreateTable("real")
	if !odb.TableExists("real") { t.Fatal("❌ real should exist") }
	fmt.Printf("✅ TEST 9 — odb.TableExists: ghost=false, real=true\n")
}

// TEST 10
func TestVersion(t *testing.T) {
	v := overdrive.Version()
	if v == "" { t.Fatal("❌ version empty") }
	if v == "unknown" { t.Fatal("❌ native lib not loaded") }
	if v != "2.3.0" { t.Fatalf("❌ expected 2.3.0, got %s", v) }
	fmt.Printf("✅ TEST 10 — odb version: %s\n", v)
}

// Suppress unused import warning
var _ = json.Marshal

File: java/src/test/java/com/afot/overdrive/OverDriveE2ETest.java
package com.afot.overdrive;

import org.junit.jupiter.api.*;
import org.junit.jupiter.api.io.TempDir;

import java.io.File;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * OverDrive-DB Java SDK — Real End-to-End Tests
 *
 * These tests ACTUALLY create .odb files using the native library.
 * A passing test proves the native lib loaded and the full stack works.
 *
 * Run: mvn test -Dtest=OverDriveE2ETest
 */
@DisplayName("OverDrive-DB Java SDK — End-to-End Tests")
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class OverDriveE2ETest {

    @TempDir
    static Path tempDir;

    private String dbPath(String name) {
        return tempDir.resolve(name + ".odb").toAbsolutePath().toString();
    }

    // ─────────────────────────────────────────────────────────────
    // TEST 1: open() creates a real .odb file on disk
    // ─────────────────────────────────────────────────────────────
    @Test
    @Order(1)
    @DisplayName("open() creates a real .odb file on disk")
    void testOpenCreatesOdbFile() {
        String path = dbPath("open_creates");

        OverDrive db = assertDoesNotThrow(
            () -> OverDrive.open(path),
            "open() must succeed — native library must be loaded"
        );
        db.close();

        File f = new File(path);
        assertTrue(f.exists(), "❌ .odb file was NOT created at: " + path);
        assertTrue(f.length() > 0, "❌ .odb file is 0 bytes — engine did not initialize");
        System.out.println("✅ .odb file created (" + f.length() + " bytes): " + path);
    }

    // ─────────────────────────────────────────────────────────────
    // TEST 2: insert() + get() roundtrip
    // ─────────────────────────────────────────────────────────────
    @Test
    @Order(2)
    @DisplayName("insert() + get() roundtrip")
    void testInsertAndGetRoundtrip() throws Exception {
        String path = dbPath("crud_roundtrip");
        OverDrive db = OverDrive.open(path);

        db.createTable("users");

        String id = db.insert("users", Map.of(
            "name", "Karthikeyan",
            "role", "engineer",
            "age",  28
        ));

        assertNotNull(id, "❌ insert must return a non-null _id");
        assertFalse(id.isEmpty(), "❌ insert must return a non-empty _id");

        Map<String, Object> doc = db.get("users", id);
        assertNotNull(doc, "❌ document must exist after insert");
        assertEquals("Karthikeyan", doc.get("name"), "❌ name mismatch");
        assertEquals("engineer",    doc.get("role"), "❌ role mismatch");

        System.out.println("✅ Inserted _id=" + id + ", name=" + doc.get("name"));
        db.close();
    }

    // ─────────────────────────────────────────────────────────────
    // TEST 3: count() is accurate
    // ─────────────────────────────────────────────────────────────
    @Test
    @Order(3)
    @DisplayName("count() returns correct number")
    void testCountIsAccurate() throws Exception {
        String path = dbPath("count_check");
        OverDrive db = OverDrive.open(path);

        db.createTable("items");

        long before = db.count("items");
        assertEquals(0, before, "❌ empty table count must be 0");

        db.insert("items", Map.of("name", "A"));
        db.insert("items", Map.of("name", "B"));
        db.insert("items", Map.of("name", "C"));

        long after = db.count("items");
        assertEquals(3, after, "❌ count must be 3 after 3 inserts");
        System.out.println("✅ count=" + after);
        db.close();
    }

    // ─────────────────────────────────────────────────────────────
    // TEST 4: SQL query with WHERE returns filtered results
    // ─────────────────────────────────────────────────────────────
    @Test
    @Order(4)
    @DisplayName("query() with WHERE clause filters correctly")
    void testSQLQueryWithWhere() throws Exception {
        String path = dbPath("sql_query");
        OverDrive db = OverDrive.open(path);

        db.createTable("products");
        db.insert("products", Map.of("name", "Apple",  "price", 10));
        db.insert("products", Map.of("name", "Banana", "price", 5));
        db.insert("products", Map.of("name", "Cherry", "price", 25));

        List<Map<String, Object>> rows = db.query("SELECT * FROM products WHERE price > 8");

        assertFalse(rows.isEmpty(), "❌ query with WHERE price>8 must return rows");

        boolean hasApple  = rows.stream().anyMatch(r -> "Apple".equals(r.get("name")));
        boolean hasCherry = rows.stream().anyMatch(r -> "Cherry".equals(r.get("name")));
        assertTrue(hasApple,  "❌ Apple (price=10) must be in results");
        assertTrue(hasCherry, "❌ Cherry (price=25) must be in results");

        System.out.printf("✅ %d rows returned%n", rows.size());
        db.close();
    }

    // ─────────────────────────────────────────────────────────────
    // TEST 5: update() changes a field — verified by get()
    // ─────────────────────────────────────────────────────────────
    @Test
    @Order(5)
    @DisplayName("update() changes a field")
    void testUpdateChangesField() throws Exception {
        String path = dbPath("update_check");
        OverDrive db = OverDrive.open(path);

        db.createTable("config");
        String id = db.insert("config", Map.of("key", "theme", "value", "light"));

        boolean updated = db.update("config", id, Map.of("value", "dark"));
        assertTrue(updated, "❌ update must return true for existing doc");

        Map<String, Object> doc = db.get("config", id);
        assertEquals("dark", doc.get("value"), "❌ value must be 'dark' after update");

        System.out.println("✅ theme updated: light → " + doc.get("value"));
        db.close();
    }

    // ─────────────────────────────────────────────────────────────
    // TEST 6: delete() removes document — count drops by 1
    // ─────────────────────────────────────────────────────────────
    @Test
    @Order(6)
    @DisplayName("delete() removes document")
    void testDeleteRemovesDocument() throws Exception {
        String path = dbPath("delete_check");
        OverDrive db = OverDrive.open(path);

        db.createTable("logs");
        String id1 = db.insert("logs", Map.of("msg", "event1"));
        String id2 = db.insert("logs", Map.of("msg", "event2"));
        assertEquals(2, db.count("logs"));

        boolean deleted = db.delete("logs", id1);
        assertTrue(deleted, "❌ delete must return true for existing doc");
        assertEquals(1, db.count("logs"), "❌ count must be 1 after delete");
        assertNull(db.get("logs", id1), "❌ deleted doc must return null");
        assertNotNull(db.get("logs", id2), "❌ other doc must still exist");

        System.out.println("✅ id1 deleted, id2 still present");
        db.close();
    }

    // ─────────────────────────────────────────────────────────────
    // TEST 7: Data persists after close() + open()
    // ─────────────────────────────────────────────────────────────
    @Test
    @Order(7)
    @DisplayName("data persists after close() + open()")
    void testDataPersistsAfterReopen() throws Exception {
        String path = dbPath("persistence");

        // Write phase
        {
            OverDrive db = OverDrive.open(path);
            db.createTable("sessions");
            db.insert("sessions", Map.of("token", "abc123", "user", "afot_admin"));
            db.sync();
            db.close();
        }

        // Read phase — fresh open
        {
            OverDrive db = OverDrive.open(path);
            long count = db.count("sessions");
            assertEquals(1, count, "❌ data must persist after close+reopen. count=" + count);

            List<Map<String, Object>> rows = db.query("SELECT * FROM sessions");
            assertEquals(1, rows.size(), "❌ must have 1 row after reopen");
            assertEquals("abc123", rows.get(0).get("token"), "❌ token must persist");

            System.out.println("✅ persisted doc: token=" + rows.get(0).get("token"));
            db.close();
        }
    }

    // ─────────────────────────────────────────────────────────────
    // TEST 8: tableExists() returns correct bool
    // ─────────────────────────────────────────────────────────────
    @Test
    @Order(8)
    @DisplayName("tableExists() returns correct bool")
    void testTableExists() throws Exception {
        String path = dbPath("table_exists");
        OverDrive db = OverDrive.open(path);

        assertFalse(db.tableExists("ghost_table"), "❌ non-existent table must return false");
        db.createTable("real_table");
        assertTrue(db.tableExists("real_table"), "❌ created table must return true");

        System.out.println("✅ ghost=false, real_table=true");
        db.close();
    }

    // ─────────────────────────────────────────────────────────────
    // TEST 9: listTables() returns created tables
    // ─────────────────────────────────────────────────────────────
    @Test
    @Order(9)
    @DisplayName("listTables() returns all created tables")
    void testListTables() throws Exception {
        String path = dbPath("list_tables");
        OverDrive db = OverDrive.open(path);

        db.createTable("alpha");
        db.createTable("beta");
        db.createTable("gamma");

        List<String> tables = db.listTables();
        assertTrue(tables.contains("alpha"), "❌ 'alpha' must appear in listTables");
        assertTrue(tables.contains("beta"),  "❌ 'beta' must appear in listTables");
        assertTrue(tables.contains("gamma"), "❌ 'gamma' must appear in listTables");

        System.out.println("✅ listTables: " + tables);
        db.close();
    }

    // ─────────────────────────────────────────────────────────────
    // TEST 10: version() is a valid string from native lib
    // ─────────────────────────────────────────────────────────────
    @Test
    @Order(10)
    @DisplayName("version() returns a valid string from native lib")
    void testVersionIsValid() {
        String v = OverDrive.version();
        assertNotNull(v, "❌ version() must not return null");
        assertFalse(v.isEmpty(), "❌ version() must return a non-empty string");
        assertNotEquals("unknown", v, "❌ version() returned 'unknown' — native lib may not have loaded");
        System.out.println("✅ SDK version: " + v);
    }
}
